Auto-deploy of content using CMS Export Build Design Doc (#17304)
department-of-veterans-affairs/va.gov-team: platform/engineering/design-docs/2020-12-10-auto-deploy-cms-build.md

Eventually, the new CMS build will replace the current GraphQL build.

#### Build script

[Metalsmith](https://github.com/segmentio/metalsmith) is the helper that puts all the pieces together. The highlights of the script that is included in the content build can be found [here](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/engineering/design-docs/2020-04-09-separate-content-build.md#build-script)

#### Drupal build script

We currently have the option of creating the build using the GraphQL (GraphQL build) or using transformers (CMS build). This is done by passing the `--use-cms-export` into the Jenkins file

##### Process to obtain the Drupal content from GraphQL

1. Metalsmith calls the `getDrupalContent` function from `/drupal/metalsmith-drupal`
2. The `getDrupalContent` function calls `loadDrupal` while passing `buildOptions` as the argument
3. The `loadDrupal` function then calls `getContentViaGraphQL`

##### Process to obtain the Drupal content from CMS

1. Metalsmith calls the `getDrupalContent` function from `/drupal/metalsmith-drupal`
2. The `getDrupalContent` function calls `loadDrupal` while passing `buildOptions` as the argument

> 3. The `loadDrupal` function verifies that the `--use-cms-export` flag is set to `true`, if so, it calls `getContentFromExport`

Reviewer: Should we add links for these functions and files to provide some context?

3. A script to verify the CMS content

## Specifics

### Detailed Design

#### A full deploy

To deploy the CMS content build inside the current auto-deploy for `vets-website` the `--use-cms-export` flag will need to be passed to the `build` function inside `common.groovy`

#### A recurrent partial deploy (every 5 mins ideally)

In addition to the full deploy, a partial deploy needs to be fired up to create a content-only build and update the AWS bucket with the newest content.

Currently, the CMS team is doing this process manually. They have been deploying content-only as needed.

CMS team has recommended an auto-deploy with 5 mins intervals while they provide support.

The idea here is to create a Jenkins/CircleCI job to run a cron job that will trigger a partial deploy, initially, every hour from 8am-8pm EST. Depending on the success rate of the build, the intervals time will be reduced.

#### A script to verify the CMS content

Deploying the partial build every X minutes is not very proficient, therefore, there needs to be a script that verifies that the CMS content is not the same.

#### When to deploy the content-only build

There needs to be a script that will:

> - Download the CMS content (`.tar`) file and will compare it with the previous one

Reviewer: Will this be from https://prod.cms.va.gov/cms-export/content? Maybe we should add where we'll be downloading the CMS content for context. Also, will we be getting the previous export from S3?

> - Download the CMS content (`.tar`) file and will compare it with the previous one

Reviewer: Is there a way to integrate [webhooks](https://www.drupal.org/project/webhooks) in this process so that rather than needing to download both  exports and compare them, we can just get the time of the last updated published content in Drupal and use that to determine if we should build or not? This is probably more a question for the CMS team though, and probably isn't possible. Just a thought.

This document _does not_ discuss the design of the full build and deploy using CMS export data.

The intended audience are front-end (FE) and DevOps engineers on the Veteran-facing Services Platform (VSP) and the CMS teams.

### Background

#### Content build and transformers

The VA.gov website is built from the `vets-website` repo. The build outputs static HTML content, React applications, and styling.

The static content is built through a Metalsmith pipeline using data pulled from the Drupal-based CMS and the [`vagov-content` repo](https://github.com/department-of-veterans-affairs/vagov-content/).

Currently, the data from the Drupal server is fetched with a GraphQL query. However, the GraphQL-based build has limitations in maintainability and scalability, so an alternative is being developed to replace it.

This implementation, the CMS export build, involves the [use of transformers](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/products/cms-integration) to transform CMS export data generated by [Tome Sync](https://tome.fyi/docs/sub-modules/) into a content model that matches that of the GraphQL data.

> In the same manner as the GraphQL data, the transformed CMS export data would then applied to the Liquid templates to generate the HTML files in the build.

Reviewer 2: ```suggestion
In the same manner as the GraphQL data, the transformed CMS export data would then be applied to the Liquid templates to generate the HTML files in the build.
```

The VA.gov website can be deployed through the partial (or content-only) deploy or the full deploy.

The full deploy is a daily job that deploys `vets-website` to the production environment of VA.gov. It uses the latest commit of `vets-website` as of 2:00pm ET on that day and the latest content downloaded from the Drupal server. It can be manually invoked as necessary, independently of the daily schedule.

The content-only deploy builds `vets-website` with the latest content downloaded from the Drupal server and deploys to the dev and staging environments of VA.gov. This job can be triggered in one of two ways:
- The daily production deploy of `vets-website`, which will build with latest release of `vets-website`

> - The `va-cms-bot`, which will build with the latest commit on `vets-website` in response to a manual trigger on the part of the CMS team

Reviewer 2: ```suggestion
- The `va-cms-bot`, which will build with the latest commit on `vets-website` in response to a manual trigger by the CMS team or CMS users with the `Content Admin` or `Content Publisher` roles 
```

> - The `va-cms-bot`, which will build with the latest commit on `vets-website` in response to a manual trigger on the part of the CMS team

Reviewer 3: The content release (content only deploy) should build with the previous release, not the latest commit.  Otherwise we would have unreleased code going out.

Reviewer 3: I also think it would be good to spell out passing commits.   Because commits that have not passed would probably not be used in either type of deploy.

- Content writers want to quickly draft, publish, and deploy content.
  - However, `vets-website` is only deployed to production once per day.
  - The CMS team can run content-only deploys independently of the daily production deploy, but that is triggered manually.
- The content-only deploy does not include automated accessibility checks.
- In local development, engineers can encounter confusing build errors that are due to the mismatch in outdated locally cached content and updated Liquid templates.

### High Level Design

This new auto-deploy has a few main differences with the existing content-only deploy.
- Uses the CMS export to build content
- Runs on a schedule instead of being triggered by other jobs (such as the daily prod deploy)
- Deploys to production instead of only dev and staging

The schedule will run between 8am and 8pm ET. It will initially be set to run hourly as a conservative cadence. It will then be incrementally adjusted to a higher frequency as the build system allows.

To avoid deploying when content hasn't changed since the previous deploy, the job will compare the export data (e.g., using and archiving the checksums of the tarballs).

## Specifics

> ### Detailed Design

Reviewer 4: Can we add a section about how to manually trigger a content only deploy?  This will still need to occur from the Drupal.

#### Creating a new pipeline

A new pipeline (e.g., `Jenkinsfile.autodeploy`) will be created for the auto-deploy. It will be modeled after the current `Jenkinsfile.content` configuration.

#### Using the CMS export to build content

In order for the partial deploy to use the CMS export data, the [build command](https://github.com/department-of-veterans-affairs/vets-website/blob/master/jenkins/common.groovy#L189) in the `build` function in `jenkins/common.groovy` will have to include the `--use-cms-export` flag.

Keep in mind that this content-only deploy will build using the latest release instead of the latest commit of of `vets-website`. This matches the behavior of the current content-only deploy. The daily production deploy will continue deploying from the latest commit and retain the responsibility of tagging releases.

If this work is done before the full build has transitioned to using the CMS export, the inclusion of the flag will be contingent on the `contentOnly` parameter to the function.

#### Validating the build

> The implementation of the full build pipeline may introduce new stages or common helper functions that the auto-deploy pipeline can incorporate. Specifically, the full build pipeline will compare the diffs between GraphQL and CMS export builds. That validation should be done in this pipeline as well while this process is being rolled out.

Reviewer 5: @indytechcook the diff comparison will only be added to the build for the purposes of release. Once CMS Export is rolled out, the assumption is that it will be removed

#### Deploying the build

> The pipeline will archive the CMS export build in S3 at a different key than the current GraphQL builds. It should match the location where the full build archives the CMS export builds.

Reviewer 2: ```suggestion
The pipeline will archive the CMS export build in S3 at a different key than the current GraphQL builds. It should match the location of the full build archives the CMS export builds.
```

For example, it might be `s3://vetsgov-website-builds-s3-upload/cms-export/${ref}/${envName}.tar.bz2`, where `ref` is the commit hash and `envName` is the environment (one of `vagovdev`, `vagovstaging`, or `vagovprod`).

The subpath `cms-export` in the key will namespace the archived CMS export build separately from the current GraphQL build. This will enable testing and debugging of the new pipeline without intefering with the current deploy process.

#### Scheduling the deploy

The pipeline will [use a cron trigger](https://www.jenkins.io/doc/book/pipeline/syntax/#triggers) to run on the desired schedule.

> The schedule to start with will be hourly between 8am and 8pm on weekdays. Ideally, the deploy would run every 5 minutes.

Reviewer 2: ```suggestion
The initial schedule will be hourly between 8am and 8pm on weekdays. Ideally, the deploy would run every 5 minutes.
```

1. Build is stored at the proper location in S3 when the pipeline is invoked.
2. Pipeline runs on the defined schedule.
3. Deploy doesn't proceed when no new content has been published.

### Logging

On top of the existing build logs, the new content-only deploy will output logs for the new transformation processes in the CMS export build.

### Debugging

The auto-deploy pipeline can be debugged by reading its console output. The S3 bucket can be checked to confirm whether the build has been deployed.

The CMS export build will validate the input and output schemas of the transformers, so errors related to the build itself can be debugged using that output.

As with the full build, during the rollout, there will be a comparison of GraphQL and CMS export builds to report errors when discrepancies between the two are found. The script used for this comparison will have some console output as well as an output JSON file describing the discrepancies.

To determine whether a page exists in the Drupal content, we can go to [`staging.va.gov/drupal/debug/`](staging.va.gov/drupal/debug/). This page is not available in production, however.

> 

Reviewer 3: Additional concerns that ought to be baked in:
1.  Will content releases fail on broken link checking?  (please no  https://github.com/department-of-veterans-affairs/va.gov-team/pull/14092)   
2.  In the event of the failure, due to broken link of some other content issued that prevents a successful build.... what will happen?  (ex: Queue for another attempt, just wait until the next timed release,...?)

### Caveats

To be determined.

### Security Concerns

There are no new security concerns with the auto-deploy. There is no user interaction involved unless we include a manual trigger for use by internal VSP engineers, which still should not present any threat of malicious requests. Denial of service is also not a concern as builds are queued.

### Privacy Concerns

There are no new privacy concerns with the auto-deploy as user data is not involved.

### Open Questions and Risks

#### Questions

- What is the ideal schedule for the auto-deploy?
  - The pipeline will initially run hourly 8am-8pm ET on weekdays.
  - We want the highest rate of deploy that the system can sustain.

- How should we coordinate the schedule of the auto-deploy with the daily production deploy?
  - The deploys from both pipelines will likely overlap.
  - The daily production deploy is still necessary for tagging the release.
  - [The queue of content-only deploys should naturally resolve conflicts without any special handling.](https://github.com/department-of-veterans-affairs/va.gov-team/issues/18393)

> - Is there a way to get the time of the last updated published content in Drupal and use that to determine if we should build or not?

Reviewer 4: Please let me know if this is a requirement for launch and I will get an issue created and work on it ASAP!

> - Is there a way to get the time of the last updated published content in Drupal and use that to determine if we should build or not?

Reviewer 6: Is it possible to bundle this info within the tarball? Realizing this could be useful for the timestamp idea you mentioned as well.

> - Is there a way to get the time of the last updated published content in Drupal and use that to determine if we should build or not?

Reviewer 4: it would be the timestamp of the newest file in the tar file.

  - If we have access to that information, we may not need to compare the CMS export tarballs.

#### Risks

- Newly added content from CMS without associated transformer schemas could fail the build.
  - New content introduces the possibility of new discrepancies, which would fail any validations that depend on diffs between the GraphQL and CMS export builds.
  - Having more content will inflate the runtime of the content-only deploy and could potentially exceed the scheduled interval.
  - The best defense against this is accounting for null and empty fields in transformers.

### Work Estimates

- Create a new content-only deploy job for CMS export builds (5 pts)
- Run the partial deploy on a schedule (3 pts)
- Auto-deploy only new content (3 pts)

### Alternatives

Ideally, content would get deployed immediately as content writers make changes. However, the CMS export process is not suited for multiple invocations in quick succession, so we opted for a process that will essentially batch and deploy content changes as appropriate.

### Future Work

> - There are plans to transition from Jenkins to CircleCI for our build pipeline. The auto-deploy job will have to be brought over to CircleCI at that time.

Reviewer 7: Since this job requires access to the CMS drupal which is on the VA network, this job cannot be run from CircleCI, is my understanding. Operations _does_ plan to have a 'thing-doer' on the internal VA network. Another option would be to use Codebuild, which already exists as an AWS service, and can be triggered with a 'cron' job. The Ops group is currently determining what tool would be the 'thing-doer', there are a number of options but the requirement to be self-hosted is a limiter to those options. If anyone has any experience with a self-hosted 'thing-doer', I would be interested to hear about it.

> - To improve the runtime, we might want to explore incremental deploys.

Reviewer 4: 👍  ❤️

To determine whether a page exists in the Drupal content, we can go to [`staging.va.gov/drupal/debug/`](staging.va.gov/drupal/debug/). This page is not available in production, however.

### Caveats

To be determined.

### Security Concerns

There are no new security concerns with the auto-deploy. There is no user interaction involved unless we include a manual trigger for use by internal VSP engineers, which still should not present any threat of malicious requests. Denial of service is also not a concern as builds are queued.

### Privacy Concerns

There are no new privacy concerns with the auto-deploy as user data is not involved.

### Open Questions and Risks

#### Questions

- What is the ideal schedule for the auto-deploy?
  - The pipeline will initially run hourly 8am-8pm ET on weekdays.
  - We want the highest rate of deploy that the system can sustain.

- How should we coordinate the schedule of the auto-deploy with the daily production deploy?
  - The deploys from both pipelines will likely overlap.

>   - The daily production deploy is still necessary for tagging the release.

Reviewer 7: Would it be better to move the 'release' job to a separate job that doesn't then trigger the deploy?

Reviewer 6: I don't find the coupling of the release and deploy to be a detriment right now, so I'm not sure if it would be necessary for the release. But having the separation would be cleaner and should probably be considered in the move to CircleCI.

We will leverage the existing full auto-deploy job (`deploys/vets-gov-autodeploy-vets-website`) by having the content-only auto-deploy invoke it.
- The `use_latest_release` flag will be set to `true` and the `release_wait` will be set to 0.
- Since the content-only auto-deploy will be a recurring job, there are some concerns around potential conflicts with daily deploys.
- Having the new job invoke `deploys/vetsgov-autodeploy-vets-website` will allow both auto-deploys to share a queue in Jenkins and avoid conflicts.
- CMS stakeholders can still manually trigger the existing job for content-only deploys as before.

The Jenkinsfile for the content-only deploy (e.g., `ansible/Jenkinsfiles/deploys/content-only-autodeploy`) would contain this fundamental step:
```
stage('Autodeploy content') {
  steps {
    build job: 'deploys/vets-gov-autodeploy-vets-website', parameters: [
      stringParam(name: 'release_wait', value: '0'),
      booleanParam(name: 'use_latest_release', value: true),
    ], wait: true
  }
}

> ```

Reviewer 6: @dginther: Just updated this section and wanted to double check my assumptions about how Jenkins works.

Basically we want to be able to run the same `deploys/vets-gov-autodeploy-vets-website` job, but with different cron trigger and parameters.

This document _does not_ discuss the design of the full build and deploy using CMS export data.

The intended audience are front-end (FE) and DevOps engineers on the Veteran-facing Services Platform (VSP) and the CMS teams.

### Background

#### Content build and transformers

The VA.gov website is built from the `vets-website` repo. The build outputs static HTML content, React applications, and styling.

The static content is built through a Metalsmith pipeline using data pulled from the Drupal-based CMS and the [`vagov-content` repo](https://github.com/department-of-veterans-affairs/vagov-content/).

Currently, the data from the Drupal server is fetched with a GraphQL query. However, the GraphQL-based build has limitations in maintainability and scalability, so an alternative is being developed to replace it.

This implementation, the CMS export build, involves the [use of transformers](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/products/cms-integration) to transform CMS export data into a content model that matches that of the GraphQL data.

The CMS export data is a JSON static file representation of the content in the Drupal database.

> 

Reviewer 4: > representing Drupal nodes.

The CMS Export also includes non-node content including, but not limited to, paragraphs, taxonomy, and menu items.

> a tar file 

The tar is a delivery method but the actual export data is a collection of json files.

To deploy content changes at will, that "auto-deploy" job is manually triggered with the `use_latest_release` flag by the CMS team or CMS users with the `Content Admin` or `Content Publisher` roles.

#### Pain points

- Content writers want to quickly draft, publish, and deploy content.
  - However, `vets-website` is only deployed to production once per day.
  - The CMS team can run content-only deploys independently of the daily production deploy, but that is triggered manually.
- The content-only deploy does not include automated accessibility checks.
- In local development, engineers can encounter confusing build errors that are due to the mismatch in outdated locally cached content and updated Liquid templates.

### High Level Design

There will be a new content-only auto-deploy job in addition to the existing full auto-deploy.
- It will build content with CMS export data.
- It will be a recurring job that executes on a more frequent schedule than the daily prod deploy.

> The schedule will run between 8am and 8pm ET to align with CMS support hours.

Reviewer 8: The initial schedule

Reviewer 6: The hourly cadence is the initial configuration, but I don't think we've discussed changing the timeframe of 8am to 8pm to something else as we roll out.

Reviewer 5: I think Dror is just saying that the hourly cadence could, and presumably will, change as part of future work. The initial schedule is hourly, 8-8 on weekdays. Future schedule will be more frequent than hourly and the timeframe may expand.

Reviewer 6: I've captured that the hourly cadence is likely to change, though I wasn't under the impression that the timeframe could expand too. I've updated this line to reflect that possibility.
